### PR TITLE
Add new parameter tuning for SMC

### PR DESCRIPTION
**Pull request by ciguaran**
# Thank you for opening a PR!
 A few important guidelines and requirements before we can merge your PR:
 - [N/A] *If I add a new sampler*, there is an issue discussing it already;
 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 
1. Implements the possibility of tuning SMC kernels based on particles.
2. As a subcase of (1) allows for tuning a proposal distribution. 
3. As a subcase of (2) allows to tune a multivariate normal distribution. 

 - [x] There are links to *all* the relevant issues, discussions and PRs;
 See step 2 of: https://github.com/blackjax-devs/blackjax/issues/245

 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date;
 - [N/a] *If I add a new sampler** I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

Consider opening a **Draft PR** if your work is still in progress but you would like some feedback from other contributors.
@rlouf @aloctavodia

### CODE REVIEW

**Comment by ciguaran** [comment]
@junpenglao beware that this PR adds new tuning functionality, not only refactors.

**Comment by junpenglao** [comment]
I see, in that case maybe it is better to split adding the functionary with the refactoring?
I think we should think a bit more of introducing the typing of `LogProbFn` and consider the pros and cons, which might be a bit easier to do in a separate PR.

**Comment by ciguaran** [comment]
@junpenglao done. removed LogProbFn from this one.

**Comment by ciguaran** [comment]
@rlouf FYI this PR is ready for review/commenting

**Comment by rlouf** [comment]
Thank you. I've been travelling/working the whole month, will take a look whenever I have a chance.

File: blackjax/smc/parameter_tunning.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+used within smc, based on particles
+"""
+
+import jax
+import jax.numpy as jnp
+
+
+def no_tunning(mcmc_algorithm, mcmc_parameters):
+    """
+    default implementation that will not take particles
```
Comment (rlouf) [inline blackjax/smc/parameter_tunning.py]: Doctrings in Blackjax start on the same line as """ wit a capitalized letter

File: blackjax/types.py [hunk 2]
```diff
@@ -15,3 +15,6 @@
 
 #: JAX PRNGKey
 PRNGKey = prng.PRNGKeyArray
+
+LogProbFn = Callable
```
Comment (rlouf) [inline blackjax/types.py]: Is that linked to #262 ? In this case it would be better to do this in another PR that updates the entire codebase

File: blackjax/kernels.py [hunk 3]
```diff
@@ -106,9 +108,10 @@ def __new__(  # type: ignore[misc]
         mcmc_parameters: Dict,
         resampling_fn: Callable,
         mcmc_iter: int = 10,
+        with_parameter_tunning: Callable = no_tunning,
```
Comment (rlouf) [inline blackjax/kernels.py]: I am not sure I like this design. Look at the separation between sampling and adaptation for the MCMC algorithm.

File: tests/versions.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import blackjax
```
Comment (rlouf) [inline tests/versions.py]: Why is that needed?

File: blackjax/smc/particle_utils.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+"""
```
Comment (rlouf) [inline blackjax/smc/particle_utils.py:1]: `*_utils.py` is a design smell. In this case it's totally fine to duplicate `num_particles`.

**Comment by rlouf** [comment]
The current design adds a coupling between sampling and adaptation that doesn't need to be. It does exist to some extent with NUTS and the window adaptation, but [AeHMC](https://github.com/aesara-devs/aehmc) for instance has completely decoupled the act of sampling from the act of updating parameters and this is what we're headed towards for `blackjax==1.0.0`.

This is what we need to implement. On the one hand adaptation algorithms that take a set of particles and return values for the parameters, on the other hand SMC samplers that take a set of particles, parameters and returns a new set of particles.

**Comment by rlouf** [comment]
I still make it easier to decouple sampling and updating parameters in #277

**Comment by junpenglao** [comment]
Hi @ciguaran would you like to pick this up again?

**Comment by ciguaran** [comment]
@junpenglao yes. Stay tuned

**Comment by ciguaran** [comment]
Closing this since https://github.com/blackjax-devs/blackjax/pull/595 implements the functionality described
